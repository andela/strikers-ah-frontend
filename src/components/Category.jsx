/* eslint-disable no-script-url */
/* eslint-disable jsx-a11y/anchor-is-valid */
/* eslint-disable no-return-assign */
/* eslint-disable prefer-destructuring */
/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import swal from '@sweetalert/with-react';
import htmlReactParser from 'html-react-parser';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHeart, faBookmark } from '@fortawesome/free-solid-svg-icons';
import '../styles/css/category.css';
import defaultImage from '../styles/img/backgound-one.jpg';
import Layout from './common/pageLayout';
import { articles } from '../redux/actions/articleByCategory';
import Spinner from './article/Spinner';

/**
 * category article display
 */
export class Category extends Component {
  /**
   *
   * @param {*} props
   * @returns {*} props
   */
  constructor(props) {
    super(props);
    this.state = {
      currentPage: 1,
      articlePerPage: 4,
    };
    this.category = this.props.match.params.category;
    this.props.articles(this.category);
  }

  /**
   *
   *@param {*} event
   * @memberof Category
   * @returns {*} currentPage
   */
  handleClick = event => {
    this.setState({
      currentPage: Number(event.target.id),
    });
  };

  /**
   *@param {*} articlesArr
   * @memberof Category
   * @returns {*} articles
   */
  mapItems = articlesArr => {
    const elements = articlesArr;
    // eslint-disable-next-line no-restricted-globals
    const url = location.origin;
    const { currentPage, articlePerPage } = this.state;

    const indexOfLastArticle = currentPage * articlePerPage;
    const indexOfFirstArticle = indexOfLastArticle - articlePerPage;
    const currentArticle = elements.slice(
      indexOfFirstArticle,
      indexOfLastArticle,
    );
    const article = currentArticle.map(item => (
      <div className="articleCard" key={item.id}>
        <a id="cardLink" href={`${url}/article/${item.slug}`}>
          <div className="cardTitle">
            <span className="articleTitle">{item.title} </span>
            {/* <span className="articleDescription"> */}
            {htmlReactParser(item.description)}
            <div className="likeAndBookMarkDiv">
              <FontAwesomeIcon icon={faHeart} id="views" />
              {item.views} Views
              <button type="button" className="bookmark">
                <FontAwesomeIcon icon={faBookmark} />
              </button>
            </div>
          </div>
          <img
            className="articleImage"
            src={item.image === 'null' ? defaultImage : item.image}
            alt="imageName"
          />
        </a>
      </div>
    ));
    return article;
  };

  /**
   *
   *@param {*} articleArray
   * @memberof Category
   * @returns {*} article
   */
  randomArticle = articleArray => {
    const topArticle =
      articleArray[Math.floor(Math.random() * articleArray.length)];
    return topArticle;
  };

  /**
   * @returns {*} render
   */
  render() {
    if (this.props.articleByCategory.Articles !== undefined) {
      const Articles = this.props.articleByCategory.Articles;
      if (Articles.error) {
        return (window.location.href = '/notfound');
      }
      if (Articles.length === 0) {
        return (
          swal(`Sorry! No Article found in ${this.category} category `),
          setTimeout(() => {
            return (window.location.href = '/topic/Other');
          }, 1000)
        );
      }

      const pageNumbers = [];
      for (
        let i = 1;
        i <= Math.ceil(Articles.length / this.state.articlePerPage);
        i += 1
      ) {
        pageNumbers.push(i);
      }

      const renderPageNumbers = pageNumbers.map(number => {
        return (
          <a
            key={number}
            id={number}
            href="javascript:;"
            onClick={this.handleClick}
          >
            {number}
          </a>
        );
      });

      const randArticle = this.randomArticle(Articles);
      return (
        <Layout>
          <div className="container">
            <div className="topArticle">
              <img
                src={
                  randArticle.image === 'null'
                    ? defaultImage
                    : randArticle.image
                }
                alt="randArticletitle"
                className="topArticleImage"
              />
              <div className="titleDiv">
                <span className="title">{randArticle.title}</span>
                {/* <p className='owner'>{randArticle.description}</p> */}
                <span className="description">
                  {htmlReactParser(randArticle.description)}
                </span>
              </div>
            </div>
            {/* <div className='popular'>
                        here are the popular
                    </div> */}
            <div className="articles">
              <p className="categoryTitle">{this.category}</p>
            </div>
            <div className="cardContainer">{this.mapItems(Articles)}</div>
            <div className="center">
              <div className="pagination">{renderPageNumbers}</div>
            </div>
          </div>
        </Layout>
      );
    }

    return (
      <div>
        <Spinner />
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  articleByCategory: state.articleByCategory,
});

export default connect(
  mapStateToProps,
  { articles },
)(Category);
