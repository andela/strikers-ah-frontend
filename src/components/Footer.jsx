/* eslint-disable react/prefer-stateless-function */
import React, { Component } from 'react';
// import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
// import { faCopyright } from '@fortawesome/free-solid-svg-icons';
import '../styles/css/footer.css';

/**
 * footer
 */
class Footer extends Component {
  /**
   * @returns {*} render
   */
  render() {
    return (
      <div className="footer">
        <p className="copyright">
          Authors Haven All Rights Reserved &#169; 2019
        </p>
      </div>
    );
  }
}

export default Footer;
