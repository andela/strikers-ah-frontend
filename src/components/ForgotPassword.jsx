/* eslint-disable react/destructuring-assignment */
/* eslint-disable class-methods-use-this */
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import InputForm from './common/InputForm';
import logo from '../styles/img/logo.png';
import { forgotPassword } from '../redux/actions/userAction';
import SubmitButton from './common/SubmitButton';

/**
 * Forgot Password Form
 */
class ForgotPassword extends Component {
  /**
   *
   * @param {*} props
   */
  constructor(props) {
    super(props);
    this.state = { value: '' };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  /**
   * @author frank harerimana
   * @param {*} e
   * @returns {*} action
   */
  handleChange(e) {
    this.setState({ value: e.target.value });
  }

  /**
   * @author frank harerimana
   * @param {*} e
   * @returns {*} action
   */
  handleSubmit(e) {
    e.preventDefault();
    // console.log(this.state.value);
    const email = this.state.value;
    this.props.forgotPassword(email);
  }

  /**
   * @author frank
   * @returns {*} render
   */
  render() {
    let message = '';
    if (this.props.forgetPasswordState) {
      const { response } = this.props.forgetPasswordState;
      message = response ? response.message : '';
    }
    return (
      <div className="container">
        <div className="login-section">
          <div className="logo">
            <img src={logo} alt="authors haven" />
          </div>
          <div className="log-text">
            <p>{message}</p>
            <form onSubmit={this.handleSubmit}>
              <InputForm
                type="email"
                name="email"
                placeholder="Email or Username"
                required
                onChange={this.handleChange}
                value={this.state.value}
              />
              <Link to="/login" className="fo-btn">
                Back To Login
              </Link>
              <SubmitButton
                className="login-btn"
                type="submit"
                value="Forgot Password"
              />
            </form>
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  forgetPasswordState: state.forgotPassword,
});

export default connect(
  mapStateToProps,
  { forgotPassword },
)(ForgotPassword);
