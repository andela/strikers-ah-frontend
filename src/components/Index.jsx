/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import queryString from 'query-string';
import { connect } from 'react-redux';
import { getLoggedInUserProfile } from '../redux/actions/userAction';

/**
 * @author frank harerimana
 */
class Index extends Component {
  /**
   * @author frank
   * @returns {*} state
   */
  componentWillMount() {
    const user = localStorage.getItem('token');
    const url = this.props.location.search;
    const params = queryString.parse(url);
    if (params.token != null) {
      localStorage.setItem('token', params.token);
      this.props.getLoggedInUserProfile(params.token);
    } else if (user != null) {
      this.props.getLoggedInUserProfile(user);
    }
  }

  /**
   * @author frank harerimana
   * @returns {*} render
   */
  render() {
    if (this.props.userProfile.length !== 0) {
      const { user } = this.props.userProfile;
      console.log(user);
    }
    return (
      <div>
        <p>home page component</p>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  userProfile: state.userProfile,
});

export default connect(
  mapStateToProps,
  { getLoggedInUserProfile },
)(Index);
