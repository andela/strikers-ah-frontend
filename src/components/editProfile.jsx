/* eslint-disable valid-jsdoc */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import Input from './common/textInput';
import TextArea from './common/textArea';
import { handleInputValueChange, addImage } from '../redux/actions/formActions';
import { editLoggedInUserProfile } from '../redux/actions/userAction';
import Button from './common/SubmitButton';
import '../styles/css/profileForm.css';

/**
 * @author Mwibutsa Floribert
 */
class EditProfile extends Component {
  /**
   * @author Mwibutsa Floribert
   * @returns { * } --
   */
  componentDidMount() {
    const {
      handleInputValueChange: handleInputChange,
      currentValues,
    } = this.props;
    handleInputChange(currentValues);
  }

  renderInput = ({ type, name, value, label, ...rest }) => {
    return (
      <Input
        name={name}
        type={type}
        value={value}
        label={label}
        handleChange={this.handleChange}
        {...rest}
      />
    );
  };

  /**
   * @param { Object } e -- event Object
   * @returns { * } --
   */
  handleSubmit = e => {
    e.preventDefault();
    const {
      editLoggedInUserProfile: editProfile,
      toggleEditProfile,
    } = this.props;
    const { values } = this.props;
    editProfile({ ...values });
    toggleEditProfile();
  };

  handleChange = e => {
    const { handleInputValueChange: handleInputChange } = this.props;
    handleInputChange({ [e.target.name]: e.target.value });
  };

  /**
   * @author Mwibutsa Floribert
   * @returns { Component } --
   */
  render() {
    const { values } = this.props;
    const { firstname, lastname, username, email, bio } = values;
    return (
      <div className="form-container" test-data="profileFormComponent">
        <form onSubmit={this.handleSubmit} className="edit-profile-form">
          <div className="input-grid">
            {this.renderInput({
              type: 'text',
              name: 'firstname',
              value: firstname,
              label: 'Firstname',
              required: 'required',
            })}
            {this.renderInput({
              type: 'text',
              name: 'lastname',
              value: lastname,
              label: 'Lastname',
              required: 'required',
            })}
          </div>
          <div className="input-grid">
            {this.renderInput({
              type: 'text',
              name: 'username',
              value: username,
              label: 'Username',
              ...(username.length
                ? { disabled: 'disabled' }
                : { required: 'required' }),
            })}
            {this.renderInput({
              type: 'email',
              name: 'email',
              value: email,
              label: 'Email',
              ...(email.length
                ? { disabled: 'disabled' }
                : { required: 'required' }),
            })}
          </div>
          <TextArea
            name="bio"
            value={bio}
            handleChange={this.handleChange}
            placeholder="Add a bio to your profile"
            required
          />
          <Button label="Save" type="submit" value="Save" />
        </form>
      </div>
    );
  }
}
const mapStateToProps = state => {
  return {
    values: state.profileForm.values,
    erros: state.profileForm.errors,
  };
};
export default connect(
  mapStateToProps,
  { handleInputValueChange, editLoggedInUserProfile, addImage },
)(EditProfile);
