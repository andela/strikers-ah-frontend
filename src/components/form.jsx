import React from 'react';
import Joi from 'joi-browser';
import axios from 'axios';
import { Link } from 'react-router-dom';
import dotenv from 'dotenv';
import Input from './common/input';
import FormFunctions from './common/formFunctions';
import '../styles/css/signup-style/signup-style.css';

dotenv.config();

/**
 * @author Clet Mwunguzi
 * @returns {*} form component
 */
class Form extends FormFunctions {
  schema = {
    firstname: Joi.string()
      .trim()
      .min(2)
      .max(15)
      .required(),
    lastname: Joi.string()
      .trim()
      .min(2)
      .max(15)
      .required(),
    username: Joi.string()
      .alphanum()
      .trim()
      .min(3)
      .max(10)
      .required(),
    email: Joi.string()
      .trim()
      .email()
      .required(),
    password: Joi.string()
      .regex(/^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%^&*])(?=.{8,20})/)
      .required(),
  };

  submit = async () => {
    // console.log('props', this.props.isSubmitted);
    const { data: userInfo } = this.props;
    const URL = `${process.env.REACT_APP_BACKEND}/api/auth/signup`;

    try {
      const { data } = await axios.post(URL, userInfo);
      if (data) {
        this.props.isSubmitted();
      }
    } catch (result) {
      if (!result.response.data) {
        const resp = { error: `${result}` };
        return this.props.stateChange({ errors: resp });
      }
      const obj = {};
      const { error } = result.response.data;
      const splitError = error.split(' ')[0];
      if (splitError === 'email' || splitError === 'username') {
        obj[splitError] = error;
        this.props.stateChange({ errors: obj });
      } else {
        this.props.stateChange({ errors: result.response.data });
      }
    }
  };

  /**
   * @author Clet Mwunguzi
   * @returns {*} input fields
   */
  render() {
    const { errors } = this.props;

    return (
      <form className="signup_form_style" onSubmit={this.handleSubmit}>
        <Input
          labelStyle="text_inputs"
          name="firstname"
          inputStyle="textboxStyle"
          label="First Name"
          borderStyle="border"
          innerLabelStyle="label"
          onChange={this.handleChange}
          error={errors}
          errorStyle="errorStyle"
        />

        <br />

        <Input
          labelStyle="text_inputs"
          name="lastname"
          inputStyle="textboxStyle"
          label="Last Name"
          borderStyle="border"
          innerLabelStyle="label"
          onChange={this.handleChange}
          error={errors}
          errorStyle="errorStyle"
        />

        <br />

        <Input
          labelStyle="text_inputs"
          name="username"
          inputStyle="textboxStyle"
          label="User Name"
          borderStyle="border"
          innerLabelStyle="label"
          onChange={this.handleChange}
          error={errors}
          errorStyle="errorStyle"
        />

        <br />

        <Input
          labelStyle="text_inputs"
          name="email"
          inputStyle="textboxStyle"
          label="Email"
          borderStyle="border"
          innerLabelStyle="label"
          onChange={this.handleChange}
          error={errors}
          errorStyle="errorStyle"
        />

        <br />

        <Input
          labelStyle="text_inputs"
          name="password"
          type="password"
          inputStyle="textboxStyle"
          label="Password"
          borderStyle="border"
          onChange={this.handleChange}
          innerLabelStyle="label"
          error={errors}
          errorStyle="errorStyle"
        />

        <br />
        <span className="errorStyle AfterSubmitError">{errors.error}</span>
        <input type="submit" value="Sign Up" className="button_style" />
        <div className="login-action-links">
          <Link to="/login">
            Already have an account? <span>Sign In</span>
          </Link>
        </div>
      </form>
    );
  }
}

export default Form;
