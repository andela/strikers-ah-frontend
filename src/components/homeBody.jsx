/* eslint-disable react/destructuring-assignment */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import '../css/Home-styles/home-body-content-style.css';
import nature from '../img/traveling.jpg';
import { main, featured, latest } from '../redux/actions/HomeAction';

/**
 * @author Clet Mwunguzi
 *
 */
class HomeBody extends Component {
  /**
   * @author Clet Mwunguzi
   * @param {*} props
   * @returns {*} Home body content
   */

  /**
   * @author Clet Mwunguzi
   * @param {*} props
   * @returns {*} latest action
   */
  componentDidMount() {
    this.props.main();
    this.props.latest();
    this.props.featured();
  }

  /**
   * @author Clet Mwunguzi
   * @returns {*} formated date
   */

  handleDateDisplay = element => {
    const month = new Date(element.createdAt).toDateString().split(' ')[1];
    const day = new Date(element.createdAt).toDateString().split(' ')[2];
    const date = `${month} ${day}`;
    return date;
  };

  /**
   * @author Clet Mwunguzi
   * @returns {*} Home body content
   */
  render() {
    let articles;
    if (this.props.homePageReducer) {
      const { homePageReducer } = this.props;
      articles = _.take(homePageReducer, 4);
    }
    let latestArticles;
    if (this.props.latestReducer) {
      const { latestReducer } = this.props;
      latestArticles = latestReducer.data;
    }

    let featuredArticles;
    if (this.props.featuredReducer) {
      const { featuredReducer } = this.props;
      featuredArticles = featuredReducer.data;
    }
    return (
      <div className="container-1">
        <div className="other-articles">
          <div className="reading-history grid-element">
            <div className="article-header">From your Reading History</div>

            {articles &&
              articles.map(element => (
                <a
                  className="article-link"
                  key={element.id}
                  href={`article/${element.slug}`}
                >
                  <div className="article">
                    <div className="article-element">
                      <img
                        src={element.image === 'null' ? nature : element.image}
                        alt="article"
                        className="article-picture"
                      />
                    </div>
                    <div className="article-content article-element">
                      <div className="article-head">{element.title}</div>
                      <div className="article-mark">
                        <span className="author-name">
                          {element.user.username}
                        </span>
                        <span className="article-time time-style">{`| ${this.handleDateDisplay(
                          element,
                        )}`}</span>
                        <span className="read-time time-style">
                          {' '}
                          . 4 min Read
                        </span>
                      </div>
                      <div className="article-excerpt time-style">
                        {element.description}
                      </div>
                    </div>
                  </div>
                </a>
              ))}
          </div>
          <div className="featured grid-element">
            <div className="article-header">
              <span className="side-header">Featured </span>
              <span className="side-link">SEE ALL</span>
            </div>
            <hr className="h-line" />
            {featuredArticles &&
              featuredArticles.map(element => (
                <a
                  className="article-link"
                  key={element.id}
                  href={`article/${element.slug}`}
                >
                  <div className="side-article">
                    <div className="side-article-head">{element.title}</div>
                    <div className="side-article-mark">
                      <span className="author-name side-author-name">
                        {element.username}
                      </span>
                      <span className="article-time time-style side-time-style">
                        {this.handleDateDisplay(element)}
                      </span>
                      <span className="read-time time-style side-time-style">
                        {' '}
                        . 4 min Read
                      </span>
                    </div>
                  </div>
                </a>
              ))}
          </div>
          <div className="latest grid-element">
            <div className="article-header">
              <span className="side-header">Latest</span>
              <span className="side-link">SEE ALL</span>
            </div>
            <hr className="h-line" />
            {latestArticles &&
              latestArticles.map(element => (
                <a
                  className="article-link"
                  key={element.id}
                  href={`article/${element.slug}`}
                >
                  <div className="side-article">
                    <div className="side-article-head">{element.title}</div>
                    <div className="side-article-mark">
                      <span className="author-name side-author-name">
                        {element.username}
                      </span>
                      <span className="article-time time-style side-time-style">
                        {this.handleDateDisplay(element)}
                      </span>
                      <span className="read-time time-style side-time-style">
                        {' '}
                        . 4 min Read
                      </span>
                    </div>
                  </div>
                </a>
              ))}
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  homePageReducer: state.homePageReducer.main,
  latestReducer: state.homePageReducer.latest,
  featuredReducer: state.homePageReducer.featured,
});
export default connect(
  mapStateToProps,
  { main, featured, latest },
)(HomeBody);
