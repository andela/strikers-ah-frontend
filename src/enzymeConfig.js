import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-localstorage-mock';

export default Enzyme.configure({
  adapter: new Adapter(),
});
