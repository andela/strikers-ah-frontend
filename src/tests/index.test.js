/* eslint-disable react/jsx-filename-extension */
/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
// import localStorage from '../../utils/browserMock';
import '../enzymeConfig';
import { Index } from '../components/Index';

const props = {
  location: {
    search: '',
  },
  getSocialUser: jest.fn(),
  socialLoginUser: {
    LoggedInUser: {
      username: 'Doe',
    },
  },
  localStorage: {
    getItem: jest.fn(),
  },
};
let wrapper;

describe('test Home page', () => {
  it('should test if the component renders when both localstorage token and query parameters are not provided', () => {
    localStorage.setItem('token', '');
    wrapper = shallow(<Index {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should test if a there was no user already logged in', () => {
    localStorage.setItem(
      'token',
      'bdkfjhwhiwue2321djfbksdhkndkfjhwmcnksjdmbndkkjbdk',
    );
    wrapper = shallow(<Index {...props} />);
    expect(props.getSocialUser).toHaveBeenCalledWith(
      localStorage.getItem('token'),
    );
  });

  it('Should test if there is a user logged in using social account', () => {
    props.location.search = '?token=jhdjhsdjhgjhsdg';
    wrapper = shallow(<Index {...props} />);
    expect(props.getSocialUser).toBeCalledWith('jhdjhsdjhgjhsdg');
  });

  it('Should render when username is not provided', () => {
    props.socialLoginUser.LoggedInUser.username = '';
    wrapper = shallow(<Index {...props} />);
  });
});
