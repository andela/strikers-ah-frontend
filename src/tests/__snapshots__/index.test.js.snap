// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Home page should test if the component renders when both localstorage token and query parameters are not provided 1`] = `ShallowWrapper {}`;
